src/main/scala/zio/keeper/Cluster.scala

```diff
@@ -1,31 +1,355 @@
 package zio.keeper
 
-import java.net.InetAddress
+import java.math.BigInteger
+import java.util.UUID
 
-import zio.{ IO, ZIO }
+import zio._
+import zio.console.putStrLn
+import zio.keeper.Cluster.Transport.TCPTransport
+import zio.keeper.Cluster.{ Credentials, Discovery }
+import zio.keeper.EchoServerProtocol.EchoMessage
+import zio.keeper.Error.{ HandshakeError, SendError }
+import zio.keeper.Error.ProtocolDef.NodeUnknown
+import zio.nio._
+import zio.nio.channels.{ AsynchronousServerSocketChannel, AsynchronousSocketChannel }
 import zio.stream.Stream
-import zio.keeper.Cluster._
 
-trait Cluster {
-  final def join: ZIO[Credentials with Discovery with Transport, Error, Unit] = ???
+import scala.reflect.ClassTag
+
+trait Cluster[A] {
+  def nodes: UIO[List[NodeId]]
+
+  def send[B <: A](data: B, receipt: NodeId): IO[Error, Unit]
+
+  def broadcast[B <: A](data: B)(implicit classTag: ClassTag[B]): IO[Error, Unit]
+
+  def receive: Stream[Error, Message[A]]
 }
 
 object Cluster {
 
+  sealed trait InternalProtocol
+
+  object InternalProtocol {
+
+    case object PleaseIntroduceYourself extends InternalProtocol
+
+    case class Iam(node: NodeId) extends InternalProtocol
+
+    case class IamNewHere(addr: InetAddress, port: Int) extends InternalProtocol
+
+    val protocolDef = new ProtocolDef[InternalProtocol] {
+      override def protocolName: String = "int_protoc"
+
+      override def version: ProtocolDef.Version =
+        ProtocolDef.Version(0, 0, 1)
+
+      override def schema =
+        message[PleaseIntroduceYourself.type](
+          (_: PleaseIntroduceYourself.type) => ZIO.succeed(Chunk.empty),
+          (_: Chunk[Byte]) => ZIO.succeed(PleaseIntroduceYourself)
+        ) <+>
+          message[Iam](
+            (msg: Iam) => ZIO.succeed(Chunk.fromArray(msg.node.value.toString.getBytes)),
+            (chunk: Chunk[Byte]) =>
+              ZIO.succeed(Iam(NodeId(UUID.fromString(new String(chunk.toArray)))))
+          ) <+>
+          message[IamNewHere](
+            (msg: IamNewHere) =>
+              ZIO.succeed(
+                Chunk.fromArray(msg.addr.address) ++ Chunk
+                  .fromArray(BigInteger.valueOf(msg.port.toLong).toByteArray)
+              ),
+            (chunk: Chunk[Byte]) =>
+              (for {
+                a   <- InetAddress.byAddress(chunk.take(4).toArray)
+                res <- ZIO.effect(new BigInteger(chunk.drop(4).toArray).intValue())
+              } yield IamNewHere(a, res)).orDie //TODO proper error handling
+          )
+
+    }
+
+    def handshake(publicAddr: InetAddress, port: Int, me: NodeId, seed: SocketAddress) =
+      for {
+        client <- ZIO.accessM[Transport] {
```

Comment:
Maybe rename `client` to `channel`, to be more explicit about its nature.

Reply:
done